batch.go

```diff
@@ -8,34 +8,31 @@ import (
 	"filippo.io/edwards25519"
 )
 
-// BatchVerifier holds entries of public keys, signature and a scalar which are used for batch verification.
+// BatchVerifier accumulates batch entries with `Add`, before performing batch verification with `Verify`.
 type BatchVerifier struct {
-	entries []ks
+	entries []entry
 }
 
-// ks represents the public key, signature and scalar which the caller wants to batch verify
-type ks struct {
+// item represents a batch entry with the public key, signature and scalar which the caller wants to verify
+type entry struct {
 	pubkey    ed25519.PublicKey
 	signature []byte
 	k         *edwards25519.Scalar
 }
 
-// NewBatchVerifier creates a Verifier that entries of signatures, keys and messages
-// can be added to for verification
+// NewBatchVerifier creates an empty `BatchVerifier`.
 func NewBatchVerifier() BatchVerifier {
 	return BatchVerifier{
-		entries: []ks{},
+		entries: []entry{},
 	}
 }
 
-// Add adds a (public key, signature, message) triple to the current batch.
-func (v *BatchVerifier) Add(publicKey ed25519.PublicKey, sig, message []byte) bool {
-	if l := len(publicKey); l != ed25519.PublicKeySize {
-		return false
-	}
-
-	if len(sig) != ed25519.SignatureSize || sig[63]&224 != 0 {
-		return false
+// Add adds a (public key, message, sig) triple to the current batch.
+//
+// Panics if `sig` is not 64 bytes long.
+func (v *BatchVerifier) Add(publicKey ed25519.PublicKey, message, sig []byte) {
+	if len(sig) != ed25519.SignatureSize {
+		panic("signature has wrong length")
```

Comment:
@FiloSottile This isn't exactly infallible (as you suggested), because the R value in the signature is used to calculate the challenge value upfront, in order to avoid a lifetime dependency where the message must live as long as the batch entry.

Actually, the only thing that's needed is the encoding of the `R` bytes, so this panic could be avoided by feeding *up to* the first 32 bytes of `sig` into the challenge hash; if `sig` is shorter than 32 bytes, then verification can fail later when we try to interpret the `R` and `s` values in `Verify`.

But I'm not sure how to do this, since `sig[:32]` will panic if `len(sig) < 32` rather than returning "up to 32 bytes".

Reply:
Assuming you don't want to zero-pad (which you could use `copy` for, but would introduce weird malleability for signatures that end in `00`), it would just look like

```
n := 32
if len(sig) < n {
    n = len(sig)
}
sig[:n]
```

but I think I like the option of leaving R nil better, as it doesn't even try to hash an invalid value.